gonum API change: undefined optimize.Local

As of `2704973b509448cfd217eeb036d9db869c80b837` in `gonum` the `optimize.Local` function has been removed.

This cause the library to fail building

`github.com/pa-m/sklearn/linear_model/Base.go:243:15: undefined: optimize.Local`

--- a/linear_model/Base.go
+++ b/linear_model/Base.go
@@ -133,7 +133,7 @@ type SGDRegressor struct {
 	LinearModel
 	Tol, Alpha, L1Ratio float
 	NJobs               int
-	Method              optimize.Method
+	Method              optimize.GlobalMethod
 }
 
 // NewSGDRegressor creates a *SGDRegressor with defaults
@@ -240,7 +240,8 @@ func (regr *SGDRegressor) Fit(X0, y0 *mat.Dense) base.Transformer {
 		// settings.Recorder = printer
 
 		method := regr.Method
-		res, err := optimize.Local(p, initialcoefs, settings, method)
+		settings.InitX = initialcoefs
+		res, err := optimize.Global(p, len(initialcoefs), settings, method)
 		unused(err)
 		//fmt.Printf("res=%s %#v\n", res.Status.String(), res)
 		// if err != nil && err.Error() != "linesearch: no change in location after Linesearcher step" {
@@ -293,7 +294,7 @@ type LinFitOptions struct {
 	L1Ratio                             float64
 	Loss                                Loss
 	Activation                          Activation
-	GOMethodCreator                     func() optimize.Method
+	GOMethodCreator                     func() optimize.GlobalMethod
 	ThetaInitializer                    func(Theta *mat.Dense)
 	Recorder                            optimize.Recorder
 	PerOutputFit                        bool
@@ -322,7 +323,7 @@ func LinFit(X, Ytrue *mat.Dense, opts *LinFitOptions) *LinFitResult {
 	nSamples, nFeatures := X.Dims()
 	_, nOutputs := Ytrue.Dims()
 	if opts.GOMethodCreator == nil && opts.Solver == "" {
-		opts.GOMethodCreator = func() optimize.Method { return &optimize.LBFGS{} }
+		opts.GOMethodCreator = func() optimize.GlobalMethod { return &optimize.LBFGS{} }
 	}
 	if opts.GOMethodCreator != nil {
 		opts.PerOutputFit = true
@@ -453,8 +454,9 @@ func LinFitGOM(X, Ytrue *mat.Dense, opts *LinFitOptions) *LinFitResult {
 	if opts.Epochs <= 0 {
 		opts.Epochs = 4e6 / nSamples
 	}
-	fSettings := func() *optimize.Settings {
+	fSettings := func(initx []float64) *optimize.Settings {
 		settings := optimize.DefaultSettings()
+		settings.InitX = initx
 		settings.Recorder = opts.Recorder
 		settings.GradientThreshold = 1e-12
 		settings.FunctionConverge = nil
@@ -498,7 +500,7 @@ func LinFitGOM(X, Ytrue *mat.Dense, opts *LinFitOptions) *LinFitResult {
 				},
 			}
 			mat.Col(thetao, o, thetaM)
-			ret, err := optimize.Local(p, thetao, fSettings(), opts.GOMethodCreator())
+			ret, err := optimize.Global(p, len(thetao), fSettings(thetao), opts.GOMethodCreator())
 			//fmt.Printf("output %d F:%v Grad:%v Status:%s\n", o, ret.F, mat.Norm(mat.NewVecDense(nFeatures, ret.Gradient), math.Inf(1)), ret.Status)
 			chanret <- fitOutputRes{o: o, ret: ret, err: err}
 		}
@@ -530,7 +532,7 @@ func LinFitGOM(X, Ytrue *mat.Dense, opts *LinFitOptions) *LinFitResult {
 				opts.Loss(Ytrue, X, mat.NewDense(nFeatures, nOutputs, theta), Ypred, Ydiff, mat.NewDense(nFeatures, nOutputs, gradSlice), opts.Alpha, opts.L1Ratio, nSamples, opts.Activation, opts.DisableRegularizationOfFirstFeature)
 			},
 		}
-		ret, err = optimize.Local(p, theta, fSettings(), opts.GOMethodCreator())
+		ret, err = optimize.Global(p, len(theta), fSettings(theta), opts.GOMethodCreator())
 		copy(theta, ret.X)
 		rmse = mat.Norm(Ydiff, 2) / float64(nOutputs)
 		epoch = ret.FuncEvaluations

--- a/linear_model/Base_test.go
+++ b/linear_model/Base_test.go
@@ -198,7 +198,7 @@ func TestGonumOptimizeRegressor(t *testing.T) {
 	bestErr := make(map[string]float)
 	bestTime := time.Second * 86400
 	bestSetup := make(map[string]string)
-	for _, method := range []optimize.Method{&optimize.GradientDescent{}, &optimize.BFGS{}, &optimize.CG{}, &optimize.LBFGS{}} {
+	for _, method := range []optimize.GlobalMethod{&optimize.GradientDescent{}, &optimize.BFGS{}, &optimize.CG{}, &optimize.LBFGS{}} {
 		for _, normalize := range []bool{false, true} {
 			testSetup := fmt.Sprintf("%T %v", method, normalize)
 			//fmt.Printf("-- TestLinearRegression normalize=%v --\n", normalize)
@@ -257,11 +257,11 @@ func TestLinFitGOM(t *testing.T) {
 	bestErr := make(map[string]float)
 	bestTime := time.Second * 86400
 	bestSetup := make(map[string]string)
-	for _, methodCreator := range []func() optimize.Method{
-		func() optimize.Method { return &optimize.GradientDescent{} },
-		func() optimize.Method { return &optimize.BFGS{} },
-		func() optimize.Method { return &optimize.CG{} },
-		func() optimize.Method { return &optimize.LBFGS{} },
+	for _, methodCreator := range []func() optimize.GlobalMethod{
+		func() optimize.GlobalMethod { return &optimize.GradientDescent{} },
+		func() optimize.GlobalMethod { return &optimize.BFGS{} },
+		func() optimize.GlobalMethod { return &optimize.CG{} },
+		func() optimize.GlobalMethod { return &optimize.LBFGS{} },
 	} {
 		for _, normalize := range []bool{false, true} {
 			testSetup := fmt.Sprintf("%T %v", methodCreator(), normalize)

--- a/linear_model/logistic.go
+++ b/linear_model/logistic.go
@@ -23,7 +23,7 @@ func NewLogisticRegression() *LogisticRegression {
 	regr.Normalize = false
 	regr.ActivationFunction = base.Logistic{}
 	regr.LossFunction = CrossEntropyLoss
-	regr.Options.GOMethodCreator = func() optimize.Method { return &optimize.LBFGS{} }
+	regr.Options.GOMethodCreator = func() optimize.GlobalMethod { return &optimize.LBFGS{} }
 	return regr
 }
 

--- a/linear_model/logistic_test.go
+++ b/linear_model/logistic_test.go
@@ -84,11 +84,11 @@ func TestLogRegExamScore(t *testing.T) {
 
 	// test Fit with various gonum/optimize Method
 
-	var GOMethodCreators = map[string]func() optimize.Method{
-		"bfgs": func() optimize.Method { return &optimize.BFGS{} },
-		"cg":   func() optimize.Method { return &optimize.CG{} },
+	var GOMethodCreators = map[string]func() optimize.GlobalMethod{
+		"bfgs": func() optimize.GlobalMethod { return &optimize.BFGS{} },
+		"cg":   func() optimize.GlobalMethod { return &optimize.CG{} },
 		//&optimize.GradientDescent{},
-		"lbfgs": func() optimize.Method { return &optimize.LBFGS{} },
+		"lbfgs": func() optimize.GlobalMethod { return &optimize.LBFGS{} },
 		//&optimize.NelderMead{},
 		//&optimize.Newton{}
 	}
@@ -222,11 +222,11 @@ func TestLogRegMicrochipTest(t *testing.T) {
 
 	// test Fit Microchip with various gonum/optimize Method
 
-	var GOMethodCreators = map[string]func() optimize.Method{
-		"bfgs": func() optimize.Method { return &optimize.BFGS{} },
-		"cg":   func() optimize.Method { return &optimize.CG{} },
+	var GOMethodCreators = map[string]func() optimize.GlobalMethod{
+		"bfgs": func() optimize.GlobalMethod { return &optimize.BFGS{} },
+		"cg":   func() optimize.GlobalMethod { return &optimize.CG{} },
 		//&optimize.GradientDescent{},
-		"lbfgs": func() optimize.Method { return &optimize.LBFGS{} },
+		"lbfgs": func() optimize.GlobalMethod { return &optimize.LBFGS{} },
 		//&optimize.NelderMead{},
 		//&optimize.Newton{}
 	}

